site/journal/2020/Autotracking Elegant DX Via Cutting-Edge CS.md

```diff
@@ -0,0 +1,548 @@
+//     markAsChanged(this, 'name')
+//     this.#name
+person.updateNameTo("Chris Krycho");
+```
+
+Exactly the same things happen if we render values or trigger changes from a Glimmer component’s template—as in the code example from the introduction:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+
+const MAX_LENGTH = 10;
+
+export default class PersonInfo extends Component {
+  @tracked name = '';
+
+  get nameLength() {
+    return this.name.length;
+  }
+
+  get remaining() {
+    return this.nameLength - MAX_LENGTH;
+  }
+
+  get showError() {
+    return this.remaining < 0;
+  }
+
+  updateNameTo = ({ target: { value } }) => this.myName = value;
+}
+```
+
+```handlebars
+<div>
+  <input {{on "input" this.updateName}} value={{this.myName}} />
+  <p class='{{if this.showError "error"}}'>
+    ({{this.nameLength}} remaining)
+  </p>
+</div>
+```
+
+Using `this.name` in the template directly evaluates `name`, which is the getter set up by `@tracked` and therefore calls `markAsUsed(this, 'name')`. Likewise, using `this.showError` and `this.nameLength` in the template evaluates those getters, which ultimately evaluate `name`, which again calls `markAsUsed(this, 'name')`. Calling `markAsUsed` tells the autotracking runtime that `this.name` is used to compute `name`, `nameLength` and `showError` in the `PersonInfo` component’s template.
+
+Triggering `updateName` by typing into the input invokes the setter for `name` installed by `@tracked`, and the setter calls `markAsChanged(this, 'name')`. Calling `markAsChanged` increments the global clock value, stores the updated clock value as the new clock value for `this.name`, and schedules a re-render.
+
+With these pieces in place, we can start to see how the system works as a whole. Reading a `@tracked` property while evaluating a value in the template informs the Glimmer VM that it was used in computing that template value. Changing a `@tracked` property bumps the global and property clock values and schedules a new render. This leads us to idea (3): using the global clock values to know when to recompute values in templates.
+
+### (3) Recomputing
+
+When rendering templates,[^reactive-again] the runtime sets up what is called a *tracking frame* for each new “computation” in the UI—values, components, helpers, modifiers, etc. A tracking frame is basically just a list of all the tracked properties that called `markAsUsed` while computing any particular value in the template. Since each tracking frame corresponds to a dynamic element of the UI, evaluating the entire UI the first time it is rendered produces a tree of tracking frames which corresponds exactly to the tree of UI components. Critically, though, a tracking frame doesn’t store the *values* of the tracked properties referenced during its computation. Instead, the frame stores only a reference to each property along with the property’s current and previous global clock values.
```

Comment:
This sentence could use a graphic explainer I think, it's difficult to capture but some thing showing the tree would really help drive how the structure that is being setup

Reply:
One option would be to have a tree of tracking frames, and a screengrab of a rendered UI element, and show the tracking frames corresponding to their portion of the UI (perhaps matching numbers/colors/both)

Reply:
Heh. I agree. I have a note written in red pen on my draft which says “graphic?” I’ll figure something out. Thanks for calling it out—good confirmation for me.